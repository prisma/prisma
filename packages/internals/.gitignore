node_modules
query-engine*
libquery_engine*
introspection-engine*
migration-engine*
prisma-fmt*
libquery*
/engines
/debug.d.ts
/debug.js
/debug.js.map
pnpm-lock.yaml
yarn-error.log
.DS_Store

