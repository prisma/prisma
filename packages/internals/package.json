{
  "name": "@prisma/internals",
  "version": "0.0.0",
  "description": "This package is intended for Prisma's internal use",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/prisma/prisma.git",
    "directory": "packages/internals"
  },
  "homepage": "https://www.prisma.io",
  "author": "Tim Suchanek <suchanek@prisma.io>",
  "bugs": "https://github.com/prisma/prisma/issues",
  "license": "Apache-2.0",
  "scripts": {
    "dev": "DEV=true node -r esbuild-register helpers/build.ts",
    "build": "node -r esbuild-register helpers/build.ts",
    "test": "jest --verbose",
    "prepublishOnly": "pnpm run build"
  },
  "files": [
    "README.md",
    "dist",
    "!**/libquery_engine*",
    "!dist/get-generators/engines/*",
    "scripts"
  ],
  "devDependencies": {
    "@swc/core": "1.2.204",
    "@swc/jest": "0.2.24",
    "@types/jest": "29.2.4",
    "@types/node": "12.20.55",
    "@types/resolve": "1.20.2",
    "esbuild": "0.15.13",
    "jest": "29.3.1",
    "jest-junit": "15.0.0",
    "mock-stdin": "1.0.0",
    "ts-node": "10.9.1",
    "typescript": "4.8.4",
    "yarn": "1.22.19"
  },
  "dependencies": {
    "@prisma/debug": "workspace:*",
    "@prisma/engine-core": "workspace:*",
    "@prisma/engines": "workspace:*",
    "@prisma/fetch-engine": "workspace:*",
    "@prisma/generator-helper": "workspace:*",
    "@prisma/get-platform": "workspace:*",
    "@prisma/prisma-fmt-wasm": "4.9.0-39.df61b5f5ab3f27359e273f1cba4e5a9c7cb8485a",
    "archiver": "5.3.1",
    "arg": "5.0.2",
    "chalk": "4.1.2",
    "checkpoint-client": "1.1.21",
    "cli-truncate": "2.1.0",
    "dotenv": "16.0.3",
    "escape-string-regexp": "4.0.0",
    "execa": "5.1.1",
    "find-up": "5.0.0",
    "fp-ts": "2.13.1",
    "fs-extra": "11.1.0",
    "fs-jetpack": "5.1.0",
    "global-dirs": "3.0.0",
    "globby": "11.1.0",
    "has-yarn": "2.1.0",
    "is-windows": "^1.0.2",
    "is-wsl": "^2.2.0",
    "new-github-issue-url": "0.2.1",
    "node-fetch": "2.6.7",
    "open": "7",
    "ora": "5.4.1",
    "p-map": "4.0.0",
    "prompts": "2.4.2",
    "read-pkg-up": "7.0.1",
    "replace-string": "3.1.0",
    "resolve": "1.22.1",
    "string-width": "4.2.3",
    "strip-ansi": "6.0.1",
    "strip-indent": "3.0.0",
    "temp-dir": "2.0.0",
    "temp-write": "4.0.0",
    "tempy": "1.0.1",
    "terminal-link": "2.1.1",
    "tmp": "0.2.1",
    "ts-pattern": "^4.0.1"
  },
  "sideEffects": false
}
