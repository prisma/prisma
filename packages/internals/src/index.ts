export { checkUnsupportedDataProxy } from './cli/checkUnsupportedDataProxy'
export { getGeneratorSuccessMessage } from './cli/getGeneratorSuccessMessage'
export {
  getPrismaConfigFromPackageJson,
  getRelativeSchemaPath,
  getSchema,
  getSchemaDir,
  getSchemaDirSync,
  getSchemaPath,
  getSchemaPathFromPackageJson,
  getSchemaPathFromPackageJsonSync,
  getSchemaPathSync,
  getSchemaSync,
} from './cli/getSchema'
export { getCLIPathHash, getProjectHash } from './cli/hashes'
export { unknownCommand } from './cli/Help'
export { HelpError } from './cli/Help'
export type {
  Command,
  Commands,
  CompiledGeneratorDefinition,
  Dictionary,
  GeneratorConfig,
  GeneratorDefinition,
  GeneratorDefinitionWithPackage,
  GeneratorFunction,
  GeneratorOptions,
} from './cli/types'
export { arg, format, isError } from './cli/utils'
export { ClientEngineType, DEFAULT_CLIENT_ENGINE_TYPE, getClientEngineType } from './client/getClientEngineType'
export { credentialsToUri, protocolToConnectorType, uriToCredentials } from './convertCredentials'
export * from './engine-commands'
export { Generator } from './Generator'
export type { ProviderAliases } from './get-generators/getGenerators'
export { getGenerator, getGenerators } from './get-generators/getGenerators'
export { getPackedPackage } from './getPackedPackage'
export { highlightDatamodel, highlightSql, highlightTS } from './highlight/highlight'
export type { IntrospectionSchemaVersion, IntrospectionWarnings } from './IntrospectionEngine'
export { IntrospectionEngine } from './IntrospectionEngine'
export * as logger from './logger'
export type { MigrateEngineLogLine } from './migrateEngineCommands'
export { canConnectToDatabase, createDatabase, dropDatabase } from './migrateEngineCommands'
export { MigrateEngineExitCode } from './migrateEngineCommands'
export { ErrorArea, isRustPanic, RustPanic } from './panic'
export type { GeneratorPaths } from './predefinedGeneratorResolvers'
export { BinaryType } from './resolveBinary'
export { engineEnvVarMap, resolveBinary } from './resolveBinary'
export { sendPanic } from './sendPanic'
export type { DatabaseCredentials } from './types'
export { assertNever } from './utils/assertNever'
export { callOnce } from './utils/callOnce'
export { canPrompt } from './utils/canPrompt'
export { drawBox } from './utils/drawBox'
export { extractPreviewFeatures } from './utils/extractPreviewFeatures'
export { formatms } from './utils/formatms'
export { formatTable } from './utils/formatTable'
export { getCommandWithExecutor } from './utils/getCommandWithExecutor'
export type { EnvPaths } from './utils/getEnvPaths'
export { getEnvPaths } from './utils/getEnvPaths'
export { version } from './utils/getVersionFromPackageJson'
export { handlePanic } from './utils/handlePanic'
export { hasOwnProperty } from './utils/hasOwnProperty'
export { isCi } from './utils/isCi'
export { isCurrentBinInstalledGlobally } from './utils/isCurrentBinInstalledGlobally'
export { isInteractive } from './utils/isInteractive'
export { isPromiseLike } from './utils/isPromiseLike'
export { jestConsoleContext, jestContext, jestProcessContext } from './utils/jestContext'
export { keyBy } from './utils/keyBy'
export { link } from './utils/link'
export { load } from './utils/load'
export { loadEnvFile } from './utils/loadEnvFile'
export { mapObjectValues } from './utils/mapObjectValues'
export { mapPreviewFeatures } from './utils/mapPreviewFeatures'
export { maskSchema } from './utils/maskSchema'
export { missingGeneratorMessage } from './utils/missingGeneratorMessage'
export { parseBinaryTargetsEnvValue, parseEnvValue } from './utils/parseEnvValue'
export { pick } from './utils/pick'
export { platformRegex } from './utils/platformRegex'
export { printConfigWarnings } from './utils/printConfigWarnings'
export { serializeQueryEngineName } from './utils/serializeQueryEngineName'
export { createSpinner } from './utils/spinner'
export type { Position } from './utils/trimBlocksFromSchema'
export { trimBlocksFromSchema, trimNewLine } from './utils/trimBlocksFromSchema'
export { tryLoadEnvs } from './utils/tryLoadEnvs'
export { warnOnce } from './warnOnce'
export * as wasm from './wasm'
export type { Platform } from '@prisma/get-platform'
export { getNodeAPIName, getPlatform } from '@prisma/get-platform'
