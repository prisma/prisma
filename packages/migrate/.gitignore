node_modules
migration-engine*
yarn-error.log
dist
build
.DS_Store
tmp/
pnpm-debug.log
write-test
*.tsbuildinfo
pnpm-lock.yaml
