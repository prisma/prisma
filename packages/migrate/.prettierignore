dist
tmp
fixtures_old