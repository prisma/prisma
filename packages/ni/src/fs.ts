import { resolve } from 'node:path'
import fs from 'node:fs'
import type { RunnerContext } from './runner'

export function getPackageJSON(ctx?: RunnerContext): any {
  const cwd = ctx?.cwd ?? process.cwd()
  const path = resolve(cwd, 'package.json')

  if (fs.existsSync(path)) {
    try {
      const raw = fs.readFileSync(path, 'utf-8')
      const data = JSON.parse(raw)
      return data
    }
    catch (e) {
      if (!ctx?.programmatic) {
        console.warn('Failed to parse package.json')
        process.exit(1)
      }

      throw e
    }
  }
}
