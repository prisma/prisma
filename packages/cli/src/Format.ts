import { arg, Command, format, formatms, formatSchema, getDMMF, HelpError } from '@prisma/internals'
import { getSchemaPathAndPrint } from '@prisma/migrate'
import chalk from 'chalk'
import fs from 'fs'

/**
 * $ prisma format
 */
export class Format implements Command {
  public static new(): Format {
    return new Format()
  }

  private static help = format(`
Format a Prisma schema.

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma format [options]

${chalk.bold('Options')}

  -h, --help   Display this help message
    --schema   Custom path to your Prisma schema

${chalk.bold('Examples')}

With an existing Prisma schema
  ${chalk.dim('$')} prisma format

Or specify a Prisma schema path
  ${chalk.dim('$')} prisma format --schema=./schema.prisma

  `)

  public async parse(argv: string[]): Promise<string | Error> {
    const before = Date.now()
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--schema': String,
      '--telemetry-information': String,
    })

    if (args instanceof Error) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    const schemaPath = await getSchemaPathAndPrint(args['--schema'])

    const output = await formatSchema({ schemaPath })

    try {
      // Validate whether the formatted output is a valid schema
      await getDMMF({
        datamodel: output,
      })
    } catch (e) {
      console.error('') // empty line for better readability
      throw e
    }

    fs.writeFileSync(schemaPath, output)
    const after = Date.now()

    return `Formatted ${chalk.underline(schemaPath)} in ${formatms(after - before)} ðŸš€`
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${Format.help}`)
    }
    return Format.help
  }
}
