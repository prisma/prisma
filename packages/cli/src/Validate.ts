import {
  arg,
  Command,
  format,
  getConfig,
  getDMMF,
  getLintWarningsAsText,
  handleLintPanic,
  HelpError,
  lintSchema,
  loadEnvFile,
  logger,
} from '@prisma/internals'
import { getSchemaPathAndPrint } from '@prisma/migrate'
import chalk from 'chalk'
import fs from 'fs'

/**
 * $ prisma validate
 */
export class Validate implements Command {
  public static new(): Validate {
    return new Validate()
  }

  private static help = format(`
Validate a Prisma schema.

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma validate [options]

${chalk.bold('Options')}

  -h, --help   Display this help message
    --schema   Custom path to your Prisma schema

${chalk.bold('Examples')}

  With an existing Prisma schema
    ${chalk.dim('$')} prisma validate

  Or specify a Prisma schema path
    ${chalk.dim('$')} prisma validate --schema=./schema.prisma

`)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--schema': String,
      '--telemetry-information': String,
    })

    if (args instanceof Error) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    loadEnvFile(args['--schema'], true)

    const schemaPath = await getSchemaPathAndPrint(args['--schema'])

    const schema = fs.readFileSync(schemaPath, 'utf-8')

    const { lintDiagnostics } = handleLintPanic(
      () => {
        // the only possible error here is a Rust panic
        const lintDiagnostics = lintSchema({ schema })
        return { lintDiagnostics }
      },
      { schema },
    )

    const lintWarnings = getLintWarningsAsText(lintDiagnostics)
    if (lintWarnings && logger.should.warn()) {
      // Output warnings to stderr
      console.warn(lintWarnings)
    }

    try {
      // Validate whether the formatted output is a valid schema
      await getDMMF({
        datamodel: schema,
      })
    } catch (e) {
      console.error('') // empty line for better readability
      throw e
    }

    // We could have a CLI flag to ignore env var validation
    await getConfig({
      datamodel: schema,
      ignoreEnvVarErrors: false,
    })

    return `The schema at ${chalk.underline(schemaPath)} is valid ðŸš€`
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${Validate.help}`)
    }
    return Validate.help
  }
}
