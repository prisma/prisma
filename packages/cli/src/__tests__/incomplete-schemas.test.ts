// describeIf is making eslint unhappy about the test names
/* eslint-disable jest/no-identical-title */

import { jestContext, serializeQueryEngineName } from '@prisma/internals'
import { DbExecute, DbPull, DbPush, MigrateDev, MigrateReset } from '@prisma/migrate'
import fs from 'fs'
import stripAnsi from 'strip-ansi'

import { Format } from '../Format'
import { Validate } from '../Validate'

const ctx = jestContext.new().assemble()

/**
 * Commands:
 * - wasm engine
 *   - format
 * - library/binary engine
 *   - validate
 *   - db push
 *   - db pull
 *   - db execute (doesn't use neither getDmmf nor getConfig directly)
 *   - migrate reset
 *   - migrate dev
 */

const dbExecuteSQLScript = `-- Drop & Create & Drop
DROP TABLE IF EXISTS 'test-dbexecute';
CREATE TABLE 'test-dbexecute' ("id" INTEGER PRIMARY KEY);
DROP TABLE 'test-dbexecute';`

function urlIsMissingValidationError(source: 'getDmmf' | 'getConfig') {
  const header = source === 'getDmmf' ? 'query-engine-NORMALIZED' : 'get-config wasm'
  return `
  Prisma schema validation - (${header})
  Error code: P1012
  error: Argument "url" is missing in data source block "db".
    -->  schema.prisma:3
     | 
   2 | 
   3 | datasource db {
   4 |   provider = "postgresql"
   5 | }
     | 
  
  Validation Error Count: 1
  [Context: ${source}]
  
  Prisma CLI Version : 0.0.0
  `
}

const envVarNotFoundValidationError = `
  Prisma schema validation - (get-config wasm)
  Error code: P1012
  error: Environment variable not found: SOME_UNDEFINED_DB.
    -->  schema.prisma:5
     | 
   4 |   provider = "postgresql"
   5 |   url      = env("SOME_UNDEFINED_DB")
     | 

  Validation Error Count: 1
  [Context: getConfig]

  Prisma CLI Version : 0.0.0
  `

const urlMustStartWithProtocolValidationError = `
 Prisma schema validation - (get-config wasm)
 Error code: P1012
 error: Error validating datasource \`db\`: the URL must start with the protocol \`postgresql://\` or \`postgres://\`.
   -->  schema.prisma:5
    | 
  4 |   provider = "postgresql"
  5 |   url      = env("SOME_DEFINED_INVALID_URL")
    | 
 
 Validation Error Count: 1
 [Context: getConfig]
 
 Prisma CLI Version : 0.0.0
   `

const aDatasourceBlockIsMissingError = `A datasource block is missing in the Prisma schema file.`
const thereIsNoDatasourceError = `
There is no datasource in the schema.


`

describe('[wasm] incomplete-schemas', () => {
  describe('datasource-block-url-env-set-invalid', () => {
    beforeEach(() => {
      ctx.fixture('incomplete-schemas/datasource-block-url-env-set-invalid/prisma')
    })

    it('format', async () => {
      const result = await Format.new().parse([])
      expect(result).toMatch(/^Formatted (.*) in \d+ms ðŸš€$/)
    })

    it('validate', async () => {
      expect.assertions(1)
      try {
        await Validate.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchSnapshot(urlMustStartWithProtocolValidationError)
      }
    })

    it('db push', async () => {
      expect.assertions(1)
      try {
        await DbPush.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchSnapshot(urlMustStartWithProtocolValidationError)
      }
    })

    it('db pull', async () => {
      expect.assertions(1)
      try {
        await DbPull.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(urlMustStartWithProtocolValidationError)
      }
    })

    it('db execute', async () => {
      fs.writeFileSync('script.sql', dbExecuteSQLScript)
      expect.assertions(1)

      try {
        await DbExecute.new().parse(['--file=./script.sql'])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
          P1012

          error: Error validating datasource \`db\`: the URL must start with the protocol \`postgresql://\` or \`postgres://\`.
            -->  schema.prisma:5
             | 
           4 |   provider = "postgresql"
           5 |   url      = env("SOME_DEFINED_INVALID_URL")
             | 


        `)
      }
    })

    it('migrate reset', async () => {
      expect.assertions(1)
      try {
        await MigrateReset.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(urlMustStartWithProtocolValidationError)
      }
    })

    it('migrate dev', async () => {
      expect.assertions(1)
      try {
        await MigrateDev.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(urlMustStartWithProtocolValidationError)
      }
    })
  })

  describe('datasource-block-url-env-unset', () => {
    beforeEach(() => {
      ctx.fixture('incomplete-schemas/datasource-block-url-env-unset/prisma')
    })

    it('validate', async () => {
      expect.assertions(1)
      try {
        await Validate.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(envVarNotFoundValidationError)
      }
    })

    it('db push', async () => {
      expect.assertions(1)
      try {
        await DbPush.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(envVarNotFoundValidationError)
      }
    })

    it('db pull', async () => {
      expect.assertions(1)
      try {
        await DbPull.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(envVarNotFoundValidationError)
      }
    })

    it('db execute', async () => {
      fs.writeFileSync('script.sql', dbExecuteSQLScript)
      expect.assertions(1)

      try {
        await DbExecute.new().parse(['--file=./script.sql'])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
          P1012

          error: Environment variable not found: SOME_UNDEFINED_DB.
            -->  schema.prisma:5
             | 
           4 |   provider = "postgresql"
           5 |   url      = env("SOME_UNDEFINED_DB")
             | 


        `)
      }
    })

    it('migrate reset', async () => {
      expect.assertions(1)
      try {
        await MigrateReset.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(envVarNotFoundValidationError)
      }
    })

    it('migrate dev', async () => {
      expect.assertions(1)
      try {
        await MigrateDev.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(envVarNotFoundValidationError)
      }
    })
  })

  describe('datasource-block-url-env-unset', () => {
    beforeEach(() => {
      ctx.fixture('incomplete-schemas/datasource-block-url-env-unset/prisma')
    })

    it('format', async () => {
      const result = await Format.new().parse([])
      expect(result).toMatch(/^Formatted (.*) in \d+ms ðŸš€$/)
    })
  })

  describe('datasource-block-no-url', () => {
    beforeEach(() => {
      ctx.fixture('incomplete-schemas/datasource-block-no-url/prisma')
    })

    it('format', async () => {
      expect.assertions(1)

      try {
        await Format.new().parse([])
      } catch (e) {
        expect(serializeQueryEngineName(stripAnsi(e.message))).toMatchInlineSnapshot(
          urlIsMissingValidationError('getDmmf'),
        )
      }
    })
  })

  describe('empty-schema', () => {
    beforeEach(() => {
      ctx.fixture('incomplete-schemas/empty-schema/prisma')
    })

    it('format', async () => {
      const result = await Format.new().parse([])
      expect(result).toMatch(/^Formatted (.*) in \d+ms ðŸš€$/)
    })
  })

  describe('datasource-block-no-url', () => {
    beforeEach(() => {
      ctx.fixture('incomplete-schemas/datasource-block-no-url/prisma')
    })

    it('db push', async () => {
      expect.assertions(1)
      try {
        await DbPush.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(urlIsMissingValidationError('getConfig'))
      }
    })

    it('db pull', async () => {
      expect.assertions(1)
      try {
        await DbPull.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(urlIsMissingValidationError('getConfig'))
      }
    })

    it('db execute', async () => {
      fs.writeFileSync('script.sql', dbExecuteSQLScript)
      expect.assertions(1)

      try {
        await DbExecute.new().parse(['--file=./script.sql'])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
          P1012

          error: Argument "url" is missing in data source block "db".
            -->  schema.prisma:3
             | 
           2 | 
           3 | datasource db {
           4 |   provider = "postgresql"
           5 | }
             | 


        `)
      }
    })

    it('migrate reset', async () => {
      expect.assertions(1)
      try {
        await MigrateReset.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(urlIsMissingValidationError('getConfig'))
      }
    })

    it('migrate dev', async () => {
      expect.assertions(1)
      try {
        await MigrateDev.new().parse([])
      } catch (e) {
        expect(stripAnsi(e.message)).toMatchInlineSnapshot(urlIsMissingValidationError('getConfig'))
      }
    })
  })
})

describe('[normalized library/binary] incomplete-schemas', () => {
  describe('datasource-block-no-url', () => {
    beforeEach(() => {
      ctx.fixture('incomplete-schemas/datasource-block-no-url/prisma')
    })

    it('validate', async () => {
      expect.assertions(1)
      try {
        await Validate.new().parse([])
      } catch (e) {
        expect(serializeQueryEngineName(stripAnsi(e.message))).toMatchInlineSnapshot(
          urlIsMissingValidationError('getDmmf'),
        )
      }
    })
  })

  describe('empty-schema', () => {
    beforeEach(() => {
      ctx.fixture('incomplete-schemas/empty-schema/prisma')
    })

    it('validate', async () => {
      const result = await Validate.new().parse([])
      expect(result).toMatch(/^The schema at (.*) is valid ðŸš€$/)
    })

    it('db push', async () => {
      expect.assertions(1)
      try {
        await DbPush.new().parse([])
      } catch (e) {
        expect(serializeQueryEngineName(stripAnsi(e.message))).toMatchInlineSnapshot(aDatasourceBlockIsMissingError)
      }
    })

    it('db pull', async () => {
      expect.assertions(1)
      try {
        await DbPull.new().parse([])
      } catch (e) {
        expect(serializeQueryEngineName(stripAnsi(e.message))).toMatchInlineSnapshot(thereIsNoDatasourceError)
      }
    })

    it('db execute', async () => {
      fs.writeFileSync('script.sql', dbExecuteSQLScript)
      expect.assertions(1)

      try {
        await DbExecute.new().parse(['--file=./script.sql'])
      } catch (e) {
        expect(serializeQueryEngineName(stripAnsi(e.message))).toMatchInlineSnapshot(thereIsNoDatasourceError)
      }
    })

    it('migrate reset', async () => {
      expect.assertions(1)
      try {
        await MigrateReset.new().parse([])
      } catch (e) {
        expect(serializeQueryEngineName(stripAnsi(e.message))).toMatchInlineSnapshot(aDatasourceBlockIsMissingError)
      }
    })

    it('migrate dev', async () => {
      expect.assertions(1)
      try {
        await MigrateDev.new().parse([])
      } catch (e) {
        expect(serializeQueryEngineName(stripAnsi(e.message))).toMatchInlineSnapshot(aDatasourceBlockIsMissingError)
      }
    })
  })
})
