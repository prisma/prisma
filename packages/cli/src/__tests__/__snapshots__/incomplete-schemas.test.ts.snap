// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[wasm] incomplete-schemas datasource-block-url-env-set-invalid db push: 
 Prisma schema validation - (get-config wasm)
 Error code: P1012
 error: Error validating datasource \`db\`: the URL must start with the protocol \`postgresql://\` or \`postgres://\`.
   -->  schema.prisma:5
    | 
  4 |   provider = "postgresql"
  5 |   url      = env("SOME_DEFINED_INVALID_URL")
    | 
 
 Validation Error Count: 1
 [Context: getConfig]
 
 Prisma CLI Version : 0.0.0
    1`] = `
Prisma schema validation - (get-config wasm)
Error code: P1012
error: Error validating datasource \`db\`: the URL must start with the protocol \`postgresql://\` or \`postgres://\`.
  -->  schema.prisma:5
   | 
 4 |   provider = "postgresql"
 5 |   url      = env("SOME_DEFINED_INVALID_URL")
   | 

Validation Error Count: 1
[Context: getConfig]

Prisma CLI Version : 0.0.0
`;

exports[`[wasm] incomplete-schemas datasource-block-url-env-set-invalid validate: 
 Prisma schema validation - (get-config wasm)
 Error code: P1012
 error: Error validating datasource \`db\`: the URL must start with the protocol \`postgresql://\` or \`postgres://\`.
   -->  schema.prisma:5
    | 
  4 |   provider = "postgresql"
  5 |   url      = env("SOME_DEFINED_INVALID_URL")
    | 
 
 Validation Error Count: 1
 [Context: getConfig]
 
 Prisma CLI Version : 0.0.0
    1`] = `
Prisma schema validation - (get-config wasm)
Error code: P1012
error: Error validating datasource \`db\`: the URL must start with the protocol \`postgresql://\` or \`postgres://\`.
  -->  schema.prisma:5
   | 
 4 |   provider = "postgresql"
 5 |   url      = env("SOME_DEFINED_INVALID_URL")
   | 

Validation Error Count: 1
[Context: getConfig]

Prisma CLI Version : 0.0.0
`;
